File: Compiladores(ano passado)/compact12b/compact/java/Tabid.java
import java.util.*;

public class Tabid<T> {
	private Vector<Stack<Hashtable<String,T>>> tables; // stacks for root swap
	private Stack<Hashtable<String,T>> symbols; // one Hashtable for each bucket
	Tabid() {
		tables = new Vector<Stack<Hashtable<String,T>>>();
		namearea(); // create the first namearea
	}
	public int nameareas()	{ return tables.size(); }
	public int namearea() { // create name area and return its number
		tables.addElement(symbols = new Stack<Hashtable<String,T>>());
		push(); // insert globals bucket into the namearea
		return nameareas();
	}
	public void namearea(int tab) { // set current namearea by number
		symbols = tables.elementAt(tab);
	}
	public void push()	{ symbols.push(new Hashtable<String,T>()); }
	public void pop()	{ symbols.pop(); }
	public int depth()	{ return symbols.size(); }
	public T put(String name, T value) { // add or replace
		return symbols.peek().put(name, value);
	}
	public boolean containsKey(String name) { return get(name) != null; }
	public Collection<String> keys() {
		List<String> lst = new ArrayList<String>();
		Enumeration<String> e = symbols.peek().keys();
		while (e.hasMoreElements())
			lst.add(e.nextElement());
		return lst;
	}
	public Collection<T> values() { return symbols.peek().values(); }
	public T get(String name) {
		Hashtable<String,T> h;
		Enumeration<Hashtable<String,T>> e = symbols.elements();
		while (e.hasMoreElements())
			if ((h = e.nextElement()).containsKey(name))
				return h.get(name);
		return null;
	}
	public String toString()	{
		return "Tabid has " + tables.size() + " nameareas (" +
			tables.indexOf(symbols) + " is current) " +
			tables.toString();
	}
}

File: Compiladores(ano passado)/Projecto Ano Passado/submissao_final/at/Tabid.cpp
#include "Tabid.hpp"

int Tabid::namearea() { // create name area and return its number
	tables.push_back(symbols = new Stack);
	push(); // insert globals bucket into the namearea
	return nameareas();
}

void Tabid::pop() {
	if (symbols->size() == 0) return;
	Hashtable *bucket = (*symbols)[0];
	Hashtable::iterator iter = bucket->begin(),
			    end  = bucket->end();
	while (iter != end)
		delete iter->second;
	bucket->clear();
	symbols->pop_back();
	delete bucket;
}

Symbol *Tabid::put(String name, Symbol *value) { // add or replace
	Hashtable& bucket = *(*symbols)[0];
	Symbol *old = bucket[name];
	bucket[name] = value;
	return old;
}

Symbol *Tabid::get(String name) {
	int size = symbols->size();
	for (int i = 0; i < size; i++) {
	  Hashtable& bucket = *(*symbols)[i];
	  if (bucket.count(name) > 0)
	    return bucket[name];
	}
	return 0;
}
